src/reanimated2/animation/transformationMatrix/matrixUtils.tsx

```diff
@@ -0,0 +1,369 @@
+    [0, 0, 0, 0],
+    [0, 0, 0, 0],
+  ];
+
+  for (let x = 0; x < 4; x++) {
+    for (let y = 0; y < 4; y++) {
+      for (let i = 0; i < 4; i++) {
+        output[x][y] += a[x][i] * b[i][y];
+      }
+    }
+  }
+
+  return output;
+}
+
+function linearOperationOnMatrix<T extends AffiniteMatrixFlat | AffiniteMatrix>(
+  _a: T,
+  func: (x: number) => number
+): T {
+  'worklet';
+
+  const isFlatOnStart = !!isAffiniteMatrixFlat(_a);
+  const a: AffiniteMatrixFlat = maybeFlattenMatrix(_a);
+
+  a.forEach((_, i) => {
+    a[i] = func(a[i]);
+  });
+
+  return isFlatOnStart ? (a as T) : (unflatten(a) as T);
+}
+
+function linearOperationOnMatrices<
+  T extends AffiniteMatrixFlat | AffiniteMatrix
+>(_a: T, _b: T, operation: 'add' | 'subtract'): T {
+  'worklet';
+
+  const isFlatOnStart = !!isAffiniteMatrixFlat(_a);
+  const a: AffiniteMatrixFlat = maybeFlattenMatrix(_a);
+  const b: AffiniteMatrixFlat = maybeFlattenMatrix(_b);
+
+  a.forEach((_, i) => {
+    switch (operation) {
+      case 'subtract':
+        a[i] -= b[i];
+        break;
+      case 'add':
+        a[i] += b[i];
+        break;
+    }
+  });
+
+  return isFlatOnStart ? (a as T) : (unflatten(a) as T);
+}
+
+export function subtractMatrices<T extends AffiniteMatrixFlat | AffiniteMatrix>(
+  _a: T,
+  _b: T
+): T {
+  'worklet';
+
+  return linearOperationOnMatrices(_a, _b, 'subtract');
+}
+
+export function addMatrices<T extends AffiniteMatrixFlat | AffiniteMatrix>(
+  _a: T,
+  _b: T
+): T {
+  'worklet';
+
+  return linearOperationOnMatrices(_a, _b, 'add');
+}
+
+export function scaleMatrix<T extends AffiniteMatrixFlat | AffiniteMatrix>(
+  _a: T,
+  scalar: number
+): T {
+  'worklet';
+
+  return linearOperationOnMatrix(_a, (x) => x * scalar);
+}
+
+export function getRotationMatrix(
+  angle: number,
+  axis: 'x' | 'y' | 'z' = 'z'
+): AffiniteMatrix {
+  'worklet';
+  const cos = Math.cos(angle);
+  const sin = Math.sin(angle);
+  switch (axis) {
+    case 'z':
+      return [
+        [cos, sin, 0, 0],
+        [-sin, cos, 0, 0],
+        [0, 0, 1, 0],
+        [0, 0, 0, 1],
+      ];
+    case 'y':
+      return [
+        [cos, 0, -sin, 0],
+        [0, 1, 0, 0],
+        [sin, 0, cos, 0],
+        [0, 0, 0, 1],
+      ];
+    case 'x':
+      return [
+        [1, 0, 0, 0],
+        [0, cos, sin, 0],
+        [0, -sin, cos, 0],
+        [0, 0, 0, 1],
+      ];
+  }
+}
+
+function norm(x: number, y: number, z: number) {
+  'worklet';
+  return Math.sqrt(x * x + y * y + z * z);
+}
+
+function transposeMatrix(_m: AffiniteMatrix): AffiniteMatrix {
+  'worklet';
+  const m = flatten(_m);
+  return [
+    [m[0], m[4], m[8], m[12]],
+    [m[1], m[5], m[9], m[13]],
+    [m[2], m[6], m[10], m[14]],
+    [m[3], m[7], m[11], m[15]],
+  ];
+}
+
+function innerProduct(a: number[], b: number[]) {
+  'worklet';
+  if (!(a.length === b.length)) {
+    console.warn(
+      `Error when evaluation inner product of ${a} and ${b}, because these arrays should have equal length`
```

Comment:
```suggestion
      `Cannot calculate inner product of two vectors of different length`
```

Reply:
I would prefer including some info about the vectors, since one of them may be NaN or [], and knowing this at the beginning could speed up debugging.